app/lib/features/pins/providers/notifiers/pins_notifiers.dart

```diff
@@ -0,0 +1,67 @@
+import 'package:acter/features/home/providers/client_providers.dart';
+import 'package:acter_flutter_sdk/acter_flutter_sdk_ffi.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+class AsyncPinsNotifier extends AutoDisposeAsyncNotifier<List<ActerPin>> {
+  late Stream<void> _listener;
+
+  Future<List<ActerPin>> _getPins() async {
+    final client = ref.watch(alwaysClientProvider);
+    return (await client.pins()).toList(); // this might throw internally
+  }
+
+  @override
+  Future<List<ActerPin>> build() async {
+    final client = ref.watch(alwaysClientProvider);
+    _listener = client.subscribeStream('pins'); // stay up to date
+    _listener.forEach((e) async {
+      state = await AsyncValue.guard(() => _getPins());
+    });
+    return _getPins();
+  }
+}
+
+class AsyncPinNotifier
+    extends AutoDisposeFamilyAsyncNotifier<ActerPin, String> {
+  late Stream<void> _listener;
+
+  Future<ActerPin> _getPin() async {
+    final client = ref.watch(alwaysClientProvider);
+    try {
+      return await client.pin(arg);
+    } catch (e) {
+      return await client.waitForPin(arg, null);
+    }
```

Comment:
going for waitForPin right away is enough. it does internally the same but more efficient.

```suggestion
      return await client.waitForPin(arg, null);
```